### PR TITLE
Refactor adaptor layer to use feature tests

### PR DESCRIPTION
**Pull request by omalley**
Refactor the code base and build scripts to compile on a wider ranger of operating systems. In particular, I've built and tested on:

* CentOS 5, 6, and 7
* Debian 6, and 8
* Mac OS 10.10
* Ubuntu 12

using the native compiler and cmake. Most of the work separates out the C++ 11 and operating system features using feature tests in Cmake.

### CODE REVIEW

File: c++/include/orc/orc-config.hh.in [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+#ifdef ORC_CXX_HAS_OVERRIDE
+  #define ORC_OVERRIDE override
+#else
+  #define ORC_OVERRIDE
+#endif
+
+#ifdef ORC_CXX_HAS_UNIQUE_PTR
+  #define ORC_UNIQUE_PTR std::unique_ptr
+  #define ORC_NEED_UTILITY
```
Comment (asandryh) [inline c++/include/orc/orc-config.hh.in]: Where is ORC_NEED_UTILITY used?

**Comment by asandryh** [comment]
What happens to C09Adapter.hh and .cc files? Are they going away?

**Comment by omalley** [comment]
The C09Adapter.hh was split into orc-config.hh.in and Adaptor.hh.in. C09Adapter was moved into OrcFile.cc so that we don't have any C++ source files that have no symbols.

**Comment by omalley** [comment]
I should explain a bit more:
* c++/include/orc/orc-config.hh.in has the stuff that impacts the API for ORC. All of the defines are prefixed with ORC and C++ code is in the orc package.
* c++/src/orc/Adaptor.hh.in has the stuff that only impacts the implementation of ORC.

**Comment by asandryh** [comment]
+1

**Pull request merged**